testing.ast

jorgefandinno opened the pull request.

rkaminsk commented:
The `remove_key_from_dict_recursively` function does probably not remove locations from all possible ASTs.

```python
from typing import Any, MutableMapping, Sequence


def remove_key_recursively(value: Any, key: Any) -> None:
    """
    This function removes the given key from any mutable mappings contained in
    in the given value composed of nested sequences, mutable mappings, and
    other values.
    """
    if isinstance(value, MutableMapping):
        if key in value:
            del value[key]
        remove_key_recursively(value.values(), key)
    elif isinstance(value, Sequence):
        for element in value:
            remove_key_recursively(element, key)
```

PS: I chose `Sequence` instead of `Iterable` because I find it a more suitable in this context.

rkaminsk commented:
I typically sort imports. Since there is a tool, I added a precommit and linter to sort/check them: #37.

Please also add an `__all__` dictionary: <https://docs.python.org/3/tutorial/modules.html#importing-from-a-package>.

rkaminsk commented:
Fine with me. Just set it to ready for review if you want me to merge this.

jorgefandinno commented:
I added here file ```print.py``` at the root.

It contains a subclass ```ASTPrettyPrinter``` of ```PrettyPrinter``` and functions ```ast_pprint_activate()``` and ```ast_pprint_activate()``

The class ```ASTPrettyPrinter``` can be used instead of ```PrettyPrinter``` and allow us to allow some parameters. For instance, removing location. I don't see how to do that without subclassing ```PrettyPrinter```.

```ast_pprint_activate()```  adds our custom format to  ```PrettyPrinter```. ```ast_pprint_activate()`` removes it.
This allow other classes that use ```PrettyPrinter``` to benefit from our format. For instance, ```TestCase```.

rkaminsk commented:
If the initializer varies between python versions, it might be easier not to define it in the derived class. The only downside is that members have to be set after initialization. Alternatively, one can add arbitrary positional and keyword arguments. But this is ugly with regards to typing.

```python
class A:
    def __init__(self, x):
        print(x)

class B(A):
    prop: int

def make_b(prop):
    b = B("Hello World")
    b.prop = prop

make_b("some prop")
```

jorgefandinno commented:
> If the initializer varies between python versions, it might be easier not to define it in the derived class. The only downside is that members have to be set after initialization. Alternatively, one can add arbitrary positional and keyword arguments. But this is ugly with regards to typing.
> 
> ```python
> class A:
>     def __init__(self, x):
>         print(x)
> 
> class B(A):
>     prop: int
> 
> def make_b(prop):
>     b = B("Hello World")
>     b.prop = prop
> 
> make_b("some prop")
> ```
> 
> Actually, I have been playing a bit and to make mypy happy, the argument/keyword argument version might be the best choice because your current version defines the property as a class variable. Not sure this is intended:
> 
> ```python
> class A:
>     a: str
> 
>     def __init__(self, a: str):
>         self.a = a
> 
> class B(A):
>     b: str
> 
>     def __init__(self, b: str, *args, **kwargs):
>         super().__init__(*args, **kwargs)
>         self.b = b
> 
>     def __str__(self) -> str:
>         return f"{self.a} {self.b}"
> 
> print(B("Hello", "World"))
> ```

Please check the current implementation.
Check the format by running 
```
python print.py 
```
in the root.

jorgefandinno commented:
What do you think about starting a new line after AST name?

Tha is

```python
ast.Function(
    location=Location(
        begin=Position(
            filename='<string>',
            line=1,
            column=8),
        end=Position(
            filename='<string>',
            line=1,
            column=16)),
    name='a',
    arguments=[ast.SymbolicTerm(
                location=Location(
                    begin=Position(
                        filename='<string>',
                        line=1,
                        column=10),
                    end=Position(filename='<string>',
                                line=1,
                                column=11)),
                symbol=Function('b', [], True)),
            ast.SymbolicTerm(
                location=Location(
                    begin=Position(
                        filename='<string>',
                        line=1,
                        column=12),
                    end=Position(
                        filename='<string>',
                        line=1,
                        column=13)),
            symbol=Function('c', [], True)),
            ast.SymbolicTerm(
                location=Location(
                    begin=Position(filename='<string>',
                        line=1,
                        column=14),
                    end=Position(
                        filename='<string>',
                        line=1,
                        column=15)),
            symbol=Function('d', [], True))],
    external=0)
```
instead of
```python
ast.Function(location=Location(begin=Position(filename='<string>',
                                              line=1,
                                              column=8),
                               end=Position(filename='<string>',
                                            line=1,
                                            column=16)),
             name='a',
             arguments=[ast.SymbolicTerm(location=Location(begin=Position(filename='<string>',
                                                                          line=1,
                                                                          column=10),
                                                           end=Position(filename='<string>',
                                                                        line=1,
                                                                        column=11)),
                                         symbol=Function('b', [], True)),
                        ast.SymbolicTerm(location=Location(begin=Position(filename='<string>',
                                                                          line=1,
                                                                          column=12),
                                                           end=Position(filename='<string>',
                                                                        line=1,
                                                                        column=13)),
                                         symbol=Function('c', [], True)),
                        ast.SymbolicTerm(location=Location(begin=Position(filename='<string>',
                                                                          line=1,
                                                                          column=14),
                                                           end=Position(filename='<string>',
                                                                        line=1,
                                                                        column=15)),
                                         symbol=Function('d', [], True))],
             external=0)
```
That way, the representation is more vertical instead of that pushed to the right.

It is a different convention that the one using in ```pprint```, but the standard only uses one character makers such as ```[``` or ```{``` while we use long names such as ```ast.Function``` that push everything a lot to the right.

rkaminsk commented:
What about the last commit to have a `pprint` replacement module. It could be implemented cleaner but this is kind of a compromise to not have to copy to much code.

rkaminsk commented:
> > What about the last commit to have a `pprint` replacement module. It could be implemented cleaner but this is kind of a compromise to not have to copy to much code.
> 
> It is good.
> 
> Let me know when this is good with you and I will revise the unisttest. I think that with this we can do better there.

One thing that still comes to my mind is pretty printing the representation of symbol objects. 

Regarding the test class, I would be fine with the current implementation. It's nicer than the clingox.pprint module in that it currently does not use any private/undocumented functions. I see the main advantage of the pprint module for debugging when working with ASTs. But go ahead if you want to.

jorgefandinno commented:
Now, the new pprint has does not use undocumented functions, does it?

jorgefandinno commented:
Also, what is the issue with symbol objects?
For ```a```, I get
```python
ast.SymbolicTerm(location=Location(begin=Position(filename='<string>',
                                                  line=1,
                                                  column=8),
                                   end=Position(filename='<string>',
                                                line=1,
                                                column=9)),
                 symbol=Function('a', [], True))
```
Is not that expected?

rkaminsk commented:
> Now, the new pprint has does not use undocumented functions, does it?

It uses all the internal ones starting with `_`. There is no guarantee that those stay the same between Python releases.

> Also, what is the issue with symbol objects?

If it is long, it will be all in one line.

jorgefandinno commented:
I changed TestCase to use pformat instead of ast_to_dict. I think the result is more readable.

I also added ```autoflake``` to pre-commit. It automatically removes unused imports from the standard library.
Let me know if you like it. It also can be configured to remove unused imports from other sources, but this needs to be added because they may have side effects.

rkaminsk commented on clingox/tests/testing/test_ast.py:
```diff
@@ -154,14 +155,15 @@ def test_assertEqual(self):
+                ?     ^ ^  ^^
+                + ast.SymbolicTerm(location=LOC,
+                ?     ^^^^^^ ^^  ^
+                +                  symbol=Function('a', [], True))-             sign=0,
```
Do you know why there is no newline here?

jorgefandinno replied:
No, that is produced by the standard ```assertEqual()```

May be that has a bug?

jorgefandinno replied:
Try this code

```python
from unittest import TestCase
from clingox.testing.ast import parse_literal, parse_term
from clingox.pprint import pformat

class ASTTestCase(TestCase):
    """
    TestCase class for testing ASTs.
    """

    def test_a(self):
        lit = parse_literal("a"),
        term = parse_term("a")
        lit_repr = pformat(lit, hide_location=True)
        term_repr = pformat(term, hide_location=True)
        print(lit_repr)
        print(term_repr)
        self.assertEqual(lit_repr, term_repr)
```
It has the problem using only standard classes.
The formatted string looks fine

rkaminsk replied:
Hmm, I suspect the output would look nicer if we append a newline to the result of `pformat()` and then compare the strings.

rkaminsk replied:
Yes, it does. I'll patch it.

rkaminsk commented:
> I changed TestCase to use pformat instead of ast_to_dict. I think the result is more readable.

We can use it like this but see the last commit which adds a newline to produce a more readable diff output.

> I also added `autoflake` to pre-commit. It automatically removes unused imports from the standard library. Let me know if you like it. It also can be configured to remove unused imports from other sources, but this needs to be added because they may have side effects.

My editor already complains about unnecessary imports but if you find it useful there is no harm in adding it. Importing the clingo module should not have side effects so you could enable even the additional option.

jorgefandinno commented:
> We can use it like this but see the last commit which adds a newline to produce a more readable diff output.

Looks good.

> My editor already complains about unnecessary imports but if you find it useful there is no harm in adding it. Importing the clingo module should not have side effects so you could enable even the additional option.

Mine also complains, but I prefer to get them removed automatically than to spend a couple of seconds doing it myself.
I added clingo and clingox as packages where is safe remove.

One can configure it to remove unused variables.

rkaminsk commented:
> One can configure it to remove unused variables.

Enable anything you think you need. :smile:

jorgefandinno commented:
> > One can configure it to remove unused variables.
> 
> Enable anything you think you need. 😄

Done, also added to nox

rkaminsk commented:
Okay, I'll have one last look this some time this week and then merge.
